// @directus/api/dist/errors/codes.js
export var ErrorCode;
(function (ErrorCode) {
    ErrorCode["ContainsNullValues"] = "CONTAINS_NULL_VALUES";
    ErrorCode["ContentTooLarge"] = "CONTENT_TOO_LARGE";
    ErrorCode["Forbidden"] = "FORBIDDEN";
    ErrorCode["IllegalAssetTransformation"] = "ILLEGAL_ASSET_TRANSFORMATION";
    ErrorCode["InvalidCredentials"] = "INVALID_CREDENTIALS";
    ErrorCode["InvalidForeignKey"] = "INVALID_FOREIGN_KEY";
    ErrorCode["InvalidIp"] = "INVALID_IP";
    ErrorCode["InvalidOtp"] = "INVALID_OTP";
    ErrorCode["InvalidPayload"] = "INVALID_PAYLOAD";
    ErrorCode["InvalidProvider"] = "INVALID_PROVIDER";
    ErrorCode["InvalidProviderConfig"] = "INVALID_PROVIDER_CONFIG";
    ErrorCode["InvalidQuery"] = "INVALID_QUERY";
    ErrorCode["InvalidToken"] = "INVALID_TOKEN";
    ErrorCode["MethodNotAllowed"] = "METHOD_NOT_ALLOWED";
    ErrorCode["NotNullViolation"] = "NOT_NULL_VIOLATION";
    ErrorCode["RangeNotSatisfiable"] = "RANGE_NOT_SATISFIABLE";
    ErrorCode["RecordNotUnique"] = "RECORD_NOT_UNIQUE";
    ErrorCode["RequestsExceeded"] = "REQUESTS_EXCEEDED";
    ErrorCode["RouteNotFound"] = "ROUTE_NOT_FOUND";
    ErrorCode["ServiceUnavailable"] = "SERVICE_UNAVAILABLE";
    ErrorCode["TokenExpired"] = "TOKEN_EXPIRED";
    ErrorCode["UnexpectedResponse"] = "UNEXPECTED_RESPONSE";
    ErrorCode["UnprocessableContent"] = "UNPROCESSABLE_CONTENT";
    ErrorCode["UnsupportedMediaType"] = "UNSUPPORTED_MEDIA_TYPE";
    ErrorCode["UserSuspended"] = "USER_SUSPENDED";
    ErrorCode["ValueOutOfRange"] = "VALUE_OUT_OF_RANGE";
    ErrorCode["ValueTooLong"] = "VALUE_TOO_LONG";
})(ErrorCode || (ErrorCode = {}));
